File: Data_Structures/fenwick_xor.cpp
#include <bits/stdc++.h>
using namespace std;

template<typename T>
struct Fenwick {
    int n;
    vector<T> fw;

    Fenwick(int _n) { n = _n + 1; fw.assign(n, 0); }

    static int lsb(int i) { return i & -i; }

    // point-xor update: a[i] ^= val
    void update(int i, T val) {
        for (++i; i < n; i += lsb(i)) fw[i] ^= val;
    }

    T query(int i) const {
        T res = 0;
        for (++i; i > 0; i -= lsb(i)) res ^= fw[i];
        return res;
    }

    T range(int l, int r) const {
        return query(r) ^ query(l-1);
    }

    T get(int i) const {
        return range(i, i);
    }

    // set a[i] = val
    void set(int i, T val) {
        T cur = get(i);
        update(i, cur ^ val);
    }
};

File: datastructures/gol/fenwick/fenwick_xor.go
package fenwick

type FenwickXor[T Integer] struct {
	n  int
	fw []T
}

func NewFenwickXor[T Integer](n int) *FenwickXor[T] {
	return &FenwickXor[T]{n: n + 1, fw: make([]T, n+1)}
}

func (f *FenwickXor[T]) lsb(i int) int { return i & -i }

// point-xor update: a[i] ^= val
func (f *FenwickXor[T]) Update(i int, val T) {
	for i++; i < f.n; i += f.lsb(i) {
		f.fw[i] ^= val
	}
}

func (f *FenwickXor[T]) Query(i int) T {
	var res T
	for i++; i > 0; i -= f.lsb(i) {
		res ^= f.fw[i]
	}
	return res
}

func (f *FenwickXor[T]) Range(l, r int) T {
	return f.Query(r) ^ f.Query(l-1)
}

func (f *FenwickXor[T]) Get(i int) T {
	return f.Range(i, i)
}

// set a[i] = val
func (f *FenwickXor[T]) Set(i int, val T) {
	cur := f.Get(i)
	f.Update(i, cur^val)
}
